File: scala/parallel_collection/download_web/async_download_web.scala
import scala.util.continuations._
import scala.actors.Actor
import scala.actors.Actor._
import scala.io.Source

import java.io.InputStream
import java.io.File
import java.net.URL
import java.nio.file.{Paths, Files, Path}
import java.nio.file.StandardCopyOption._

case class URLOpen(val url: URL, val k: (InputStream => Unit))
case class CreateFile(val filePath: Path, val stream: InputStream, val k: (Path => Unit))

class URLActor extends Actor {
	def act() {
		loop {
			react {
				case uo: URLOpen => {
					println("receive : " + uo.url)
					uo.k(uo.url.openStream())
				}
				case rs: CreateFile => {
					println("file create")
					Files.copy(rs.stream, rs.filePath, REPLACE_EXISTING)
					rs.k(rs.filePath)
				}
			}
		}
	}

	def stop() {
		exit()
	}
}

val dir = args(0)

Source.stdin.getLines.toList.par.foreach {u =>
	val url = new URL(u)

	reset {
		val actor = new URLActor()
		actor.start
		println("actor.start")

		val stream = shift {k: (InputStream => Unit) =>
			actor ! URLOpen(url, k)
			println("actor ! URLOpen")
		}

		println("stream = " + stream)

		val file = shift {k: (Path => Unit) =>
			val f = new File(url.getFile()).getName()
			val filePath = Paths.get(dir, f)

			actor ! CreateFile(filePath, stream, k)
			println("actor ! CreateFile")
		}

		println("downloaded: " + file)

		actor.stop
		println("actor.stop")
	}
}

println("*** out reset")

File: java/7.0/nio2/download_web/AsyncDownloadWeb.java

import java.io.*;
import java.util.*;
import java.net.URI;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.concurrent.Callable;
import java.util.concurrent.Executors;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Future;
import static java.nio.file.StandardCopyOption.*;

public class AsyncDownloadWeb {
	public static void main(String[] args) throws Exception {
		if (args.length < 1) {
			System.out.println("java AsyncDownloadWeb <output dir>");
			return;
		}

		BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
		final ExecutorService exec = Executors.newCachedThreadPool();
		List<Future<?>> futureList = new ArrayList<>();

		final String dir = args[0];
		String urlString = null;

		while ((urlString = reader.readLine()) != null) {
			final URI uri = URI.create(urlString);

			//URL毎の並列実行
			futureList.add(exec.submit(new Runnable() {
				public void run() {
					//URL接続処理の非同期実行
					final Future<InputStream> stream = exec.submit(new Callable<InputStream>() {
						public InputStream call() throws Exception {
							return uri.toURL().openStream();
						}
					});

					//ダウンロード処理の非同期実行
					Future<Path> file = exec.submit(new Callable<Path>() {
						public Path call() throws Exception {
							String fileName = new File(uri.getPath()).getName();
							Path filePath = Paths.get(dir, fileName);

							Files.copy(stream.get(), filePath, REPLACE_EXISTING);
							return filePath;
						}
					});

					/* ダウンロード完了まで待機
					 * call 処理内で発生した例外は Future の get 時に
					 * throw される
					 */
					try {
						System.out.printf("downloaded: %s => %s\n", uri, file.get());
					} catch (Exception ex) {
						System.out.printf("failed: %s, %s\n", uri, ex);
					}
				}
			}));
		}

		/* 完了まで待機
		 *
		 * 以下を実施せずにいきなり shutdown を呼び出すと、
		 * run 処理内での submit 前に shutdown が
		 * 呼び出された場合に不都合があるため
		 */
		for (Future<?> f : futureList) {
			f.get();
		}

		//終了
		exec.shutdown();
	}
}